First tests for User Operator

scholzj opened the pull request:
### Type of change

- ~~Bugfix~~
- Enhancement / new feature
- ~~Refactoring~~
- ~~Documentation~~

### Description

This PR adds the first tests for the User Operator. In particular:
- `UserOperatorConfig` class
- `KafkaUserModel` class

As part of this change, the `MockCertManager` class has been moved to the `operator-common` test jar to be shared between CO and UO.

Further tests will be added later to this PR or in another PR depending on fast you are with reviews ;-).

tombentley commented on user-operator/pom.xml:
```diff
@@ -48,6 +48,40 @@
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+        </dependency>
+        <!--<dependency>
```
Why are we adding commented out code?

tombentley commented on user-operator/src/main/java/io/strimzi/operator/user/UserOperatorConfig.java:
```diff
@@ -64,12 +64,16 @@ public static UserOperatorConfig fromMap(Map<String, String> map) {
-            for (String label : labelsArray) {
-                String[] fields = label.split("=");
-                labels.put(fields[0].trim(), fields[1].trim());
+        try {
```
This parsing should be factored into its own method. Possibly it should be somewhere common so it can be reused with the other operators.

tombentley commented on user-operator/src/test/java/io/strimzi/operator/user/UserOperatorConfigTest.java:
```diff
@@ -0,0 +1,100 @@
+        assertEquals(envVars.get(UserOperatorConfig.STRIMZI_CA_NAMESPACE), config.getCaNamespace());
+    }
+
+    @Test(expected = IllegalArgumentException.class)
```
I wonder whether we should be using IAE here. Really the error occurs because the env var is not set (rather than it having an illegal value). Maybe we should have a common `InvalidConfigurationException` which all the operators throw when their environment isn't configured correctly.

tombentley commented on user-operator/src/test/java/io/strimzi/operator/user/model/KafkaUserModelTest.java:
```diff
@@ -0,0 +1,91 @@
+
+    @Test
+    public void testFromCrd()   {
+        KafkaUserModel model = KafkaUserModel.fromCrd(new MockCertManager(), user, clientsCa, null);
```
I would make `MockCertManager` a field (thus only instantiated once rather than _N_ times, which helps if we ever had to add ctor parameters to it).

scholzj commented:
@tombentley I added test for `KafkaUserOperator` class and addressed your comments. In particular:
* New `ConfigUtils` class is now in `operator-common` with method for parsing labels
* New `InvalidConfigurationException` is in `operator-common` and used by `UserOperatorConfig`. Using it in CO should be done in separate PR, this is already getting big even without it.

tombentley commented on operator-common/src/main/java/io/strimzi/operator/common/ConfigUtils.java:
```diff
@@ -0,0 +1,38 @@
+     * @return  Map with parsed labels
+     * @throws InvalidConfigurationException
+     */
+    public static Map<String, String> parseLabels(String stringLabels) throws InvalidConfigurationException {
```
Now I look at this method, and alluding to my earlier comment about `Labels` class, maybe we should put this there, rather than introducing a new Utils class for it?

scholzj replied:
I'm not completely sure it belongs there. It is used only for parsing of config when starting the operators. `Labels` are used all over the place. But if you want I can move it there.

If we put it to `Labels` it should probably return `Labels` and not map, right? And we should probably also remove the InvalidConfigException from there.

tombentley replied:
We pass `Labels` around, but conceptually it's just a `Map<String, String>` and that's what we always have to convert it into when we use the Fabric8 API. What `Labels` give us it an immutable representation of that Map, which prevents people doing nasty things like mutating a `Labels` they were passed. I suppose its immutability is useful enough that it's worth keeping. A `Labels.parseLabels()` should return a Labels instance. 

I don't understand your point about `InvalidConfigException`.

tombentley replied:
OIC. Yes that's right.

scholzj commented:
@tombentley Ok, it should be done now.

tombentley commented on operator-common/src/main/java/io/strimzi/operator/common/model/Labels.java:
```diff
@@ -117,6 +117,26 @@ public static Labels fromMap(Map<String, String> labels) {
+            String[] labelsArray = stringLabels.split(",");
+            for (String label : labelsArray) {
+                String[] fields = label.split("=");
+                labels.put(fields[0].trim(), fields[1].trim());
```
Rather than throwing AIOOBE this method should throw IAE if the string is not valid according to the syntax it supports. Why? AIOOBE is impl specific; it's a much better contract to have a single exception type to cover the illegal argument case
